File: backend/src/Sol.Catalog.Infrastructure/Persistence/Repositories/ProductRepository.cs
using Microsoft.EntityFrameworkCore;
using Sol.Catalog.Application.Abstractions.Persistence;
using Sol.Catalog.Application.Common;
using Sol.Catalog.Domain.Products;
using Sol.Catalog.Infrastructure.Persistence.Configurations;

namespace Sol.Catalog.Infrastructure.Persistence.Repositories;

internal sealed class ProductRepository(CatalogDbContext db) : IProductReader, IProductWriter
{
    public async Task<PagedResult<Product>> SearchAsync(
        ProductQuerySpec spec,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(spec);

        IQueryable<Product> query = db.Products.AsNoTracking();

        query = ApplyFilters(query, spec);

        int total = await query.CountAsync(cancellationToken).ConfigureAwait(false);

        if (total == 0)
        {
            return PagedResult.Empty<Product>(spec.Page, spec.PageSize);
        }

        List<Product> items = await ApplySort(query, spec)
            .Skip((spec.Page - 1) * spec.PageSize)
            .Take(spec.PageSize)
            .ToListAsync(cancellationToken)
            .ConfigureAwait(false);

        return new PagedResult<Product>(items, spec.Page, spec.PageSize, total);
    }

    public async Task<Product?> GetByIdAsync(int id, CancellationToken cancellationToken) =>
        await db.Products
            .FirstOrDefaultAsync(p => p.Id == id, cancellationToken)
            .ConfigureAwait(false);

    public void Update(Product product)
    {
        ArgumentNullException.ThrowIfNull(product);
        db.Products.Update(product);
    }

    private static IQueryable<Product> ApplyFilters(IQueryable<Product> query, ProductQuerySpec spec)
    {
        if (!string.IsNullOrWhiteSpace(spec.SearchTerm))
        {
            string pattern = $"%{SearchNormalizer.NormalizeForLike(spec.SearchTerm)}%";

            query = query.Where(p => EF.Functions.Like(
                EF.Property<string>(p, ProductConfiguration.SearchTextProperty),
                pattern,
                SearchNormalizer.LikeEscape.ToString()));
        }

        if (spec.MinPrice is { } min)
        {
            query = query.Where(p => p.Price.Amount >= min);
        }

        if (spec.MaxPrice is { } max)
        {
            query = query.Where(p => p.Price.Amount <= max);
        }

        if (!string.IsNullOrWhiteSpace(spec.Currency))
        {
            query = query.Where(p => p.Price.Currency == spec.Currency);
        }

        if (spec.OnlyInStock)
        {
            query = query.Where(p => p.Stock > 0);
        }

        return query;
    }

    private static IOrderedQueryable<Product> ApplySort(IQueryable<Product> query, ProductQuerySpec spec)
    {
        bool asc = spec.SortDirection == SortDirection.Asc;

        return spec.SortBy switch
        {
            ProductSortField.Price => asc
                ? query.OrderBy(p => p.Price.Amount).ThenBy(p => p.Id)
                : query.OrderByDescending(p => p.Price.Amount).ThenBy(p => p.Id),

            ProductSortField.Stock => asc
                ? query.OrderBy(p => p.Stock).ThenBy(p => p.Id)
                : query.OrderByDescending(p => p.Stock).ThenBy(p => p.Id),

            ProductSortField.Sku => asc
                ? query.OrderBy(p => p.Sku.Value).ThenBy(p => p.Id)
                : query.OrderByDescending(p => p.Sku.Value).ThenBy(p => p.Id),

            _ => asc
                ? query.OrderBy(p => p.Name).ThenBy(p => p.Id)
                : query.OrderByDescending(p => p.Name).ThenBy(p => p.Id),
        };
    }
}

File: frontend/lib/features/products/domain/repositories/product_repository.dart
import 'package:decimal/decimal.dart';
import 'package:equatable/equatable.dart';
import 'package:sol_catalog/features/products/domain/entities/product.dart';
import 'package:sol_catalog/features/products/domain/entities/product_page.dart';

enum ProductSortField {
  name('Name', 'Nombre'),
  price('Price', 'Precio'),
  stock('Stock', 'Stock'),
  sku('Sku', 'SKU');

  const ProductSortField(this.wire, this.label);

  final String wire;
  final String label;
}

enum SortDirection {
  asc('Asc', 'ascendente'),
  desc('Desc', 'descendente');

  const SortDirection(this.wire, this.label);

  final String wire;
  final String label;
}

class ProductQuery extends Equatable {
  const ProductQuery({
    this.search = '',
    this.page = 1,
    this.sortBy = ProductSortField.name,
    this.sortDir = SortDirection.asc,
    this.minPrice,
    this.maxPrice,
    this.currency,
    this.onlyInStock = false,
  });

  final String search;
  final int page;
  final ProductSortField sortBy;
  final SortDirection sortDir;
  final Decimal? minPrice;
  final Decimal? maxPrice;
  final String? currency;
  final bool onlyInStock;

  bool get isDefaultSort =>
      sortBy == ProductSortField.name && sortDir == SortDirection.asc;

  int get activeFilterCount => [
    minPrice != null,
    maxPrice != null,
    currency != null,
    onlyInStock,
    !isDefaultSort,
  ].where((active) => active).length;

  ProductQuery copyWith({String? search, int? page}) => ProductQuery(
    search: search ?? this.search,
    page: page ?? this.page,
    sortBy: sortBy,
    sortDir: sortDir,
    minPrice: minPrice,
    maxPrice: maxPrice,
    currency: currency,
    onlyInStock: onlyInStock,
  );

  @override
  List<Object?> get props => [
    search,
    page,
    sortBy,
    sortDir,
    minPrice,
    maxPrice,
    currency,
    onlyInStock,
  ];
}

abstract interface class ProductRepository {
  Future<ProductPage> search(ProductQuery query);

  Future<Product> updatePrice({
    required int id,
    required Decimal price,
    required String currency,
    int? expectedVersion,
  });
}
